### ISSUE TITLE
Are the rolling_median weights correct?

### ISSUE THREAD

**Issue by MarcoGorelli** [pola-rs/polars#10031]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
pl.Series([1, 2]).rolling_median(window_size=2, weights=[0, 1])
```

### Issue description

The above outputs
```python
shape: (2,)
Series: '' [f64]
[
        null
        1.0
]
```

### Expected behavior

If the weights are `[0, 1]`, shouldn't only the second element be considered, and the result be `2.`?

So, I'd have expected
```python
```python
shape: (2,)
Series: '' [f64]
[
        null
        2.0
]
```

cc @magarick in case you know 

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.8
Index type:          UInt32
Platform:            Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python:              3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           0.9.0
fsspec:              <not installed>
matplotlib:          3.7.1
numpy:               1.25.1
pandas:              2.0.3
pyarrow:             12.0.1
pydantic:            2.0.2
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>

```

</details>

**Comment by MarcoGorelli** [comment]
Also, seems like the weights don't take effect unless they're `0`?

```python
In [9]: pl.Series([1, 2]).rolling_median(window_size=2, weights=[.5, .5])
Out[9]:
shape: (2,)
Series: '' [f64]
[
        null
        1.5
]

In [10]: pl.Series([1, 2]).rolling_median(window_size=2, weights=[.0001, .9999])
Out[10]:
shape: (2,)
Series: '' [f64]
[
        null
        1.5
]

In [11]: pl.Series([1, 2]).rolling_median(window_size=2, weights=[0, 1])
Out[11]:
shape: (2,)
Series: '' [f64]
[
        null
        1.0
]
```

**Comment by MarcoGorelli** [comment]
For reference, R `weighted.quantile` gives `2`, which is more what I would have expected:
```R
> weighted.quantile(x=c(1, 2), w=c(0, 1), 0.5, FALSE)
[1] 2
> weighted.quantile(x=c(1, 2), w=c(.5, .5), 0.5, FALSE)
[1] 1.5
```

**Comment by magarick** [comment]
@MarcoGorelli Yeah, it's a bug, ugh.
Weights for quantiles are handled differently than weights for something like a mean. The code does appropriately skip zero weights in the middle but the first weight actually ends up being somewhat special. This is an easy enough fix. I'll handle it shortly. Thanks for spotting it.

**Comment by MarcoGorelli** [comment]
thanks!

can we add unit tests on the Python side please? I'm a bit concerned that the only tests for weighted quantiles are in the docstrings

**Comment by magarick** [comment]
Oh, and for posterity, the `[.0001, .9999]` and `[.5, .5]` being the same actually is correct, weirdly enough. If you think about what the weights for a quantile are being applied to, it's not actually individual numbers but the intervals between them.
To see this, consider non-negative weights (weighted quantiles with negative weights don't really make sense) associated with the sorted values. If you normalize the weights to 1, then the ith weight, $w_i$ corresponds to the jump of the ECDF at the ith value $v_i$. That is, our weights are probabilities for a discrete distribution over our values and asking about the quantile is related to asking about the ECDF.
So, if $P_i$, and $P_{i+1}$ are the i and i+1st cumulative probability (cumsum of weights) and we're asking for the quantile at $p$ between those two, we can weight between $v_i$ and $v_{i+1}$ depending on how close $q$ is to each. But what happens when $p < P_0$? Well, $v_0$ is the smallest value so it should only correspond to $q = 0$ and that means we still need to take $v_1$ into account. So the $w_0$ is more like an offset for everything.

Anyway, almost done and yes I'll add tests.

**Comment by MarcoGorelli** [comment]
thanks for explaining! I'll try to work through the references you posted in the original PR to make sense of this more

Thanks!

And do you know of any Python package which implements weighted quantiles in the same way? If so, it might be good to compare against it in parametric hypothesis tests too

**Comment by magarick** [comment]
> And do you know of any Python package which implements weighted quantiles in the same way? If so, it might be good to compare against it in parametric hypothesis tests too

xarray might but I didn't look too carefully. Other than that, I don't see much.

**Comment by MarcoGorelli** [comment]
closed by #10051 - nice one @magarick !
